### PR TITLE
Fixes #146

### LINKED ISSUE #146: [BUG] Unhandled Exception: type 'Future<ByteData?>' is not a subtype of type 'Future<ByteData>'
**Describe the bug**
Whenever I start a new app I hit some break point in VSCode and an unhandled exception gets printed to the console:

```
[VERBOSE-2:ui_dart_state.cc(186)] Unhandled Exception: type 'Future<ByteData?>' is not a subtype of type 'Future<ByteData>' in type cast
#0      _loadFontByteData package:google_fonts/src/google_fonts_base.dart:187
<asynchronous suspension>
```

**To Reproduce**
1. `flutter create blah
2. Add a `Text` widget in a `Column` in a `Center`
3. Use this:
  ```
     Text(
          "blah blah",
          style: GoogleFonts.lato(
            textStyle: TextStyle(
              color: Colors.cyan[900],
              fontSize: 16,
              wordSpacing: 3,
              letterSpacing: 7,
              fontWeight: FontWeight.bold,
            ),
          ),
        ),
  ``` 

**Expected behavior**
App doesn't throw an exception

**Include the contents of your `google_fonts` directory**
No google_fonts directory

**Screenshots**
image

**Emulator:**
 - OS: iOS
 - Version: 14.4
 - Model: 12 Pro Max

google_fonts version: "2.0.0-nullsafety.0"

### PR DESCRIPTION
**Pull request by gaetschwartz**
This PR fixes https://github.com/material-foundation/google-fonts-flutter/issues/146. 

As per discussed in #148 it would be better to tackle this issue first, so here is the PR.

WWhile running the tests, the `google-fonts-flutter\test\load_font_if_necessary_test.dart: loadFontIfNecessary method calls http get` test hangs indefinitely regardless of my fix or not. Am I doing anything wrong ?

### CODE REVIEW

**Comment by gaetschwartz** [comment]
CI is failing because of `path_provider`'s `getTemporaryDirectory` being non nullable, https://github.com/material-foundation/google-fonts-flutter/pull/148 fixes it.

**Comment by kevmoo** [comment]
@clocksmith @pennzht @johnsonmh ?

File: lib/src/google_fonts_base.dart [hunk 1]
```diff
@@ -181,12 +181,13 @@ Future<void> _loadFontByteData(
...
-    final fontLoader = FontLoader(familyWithVariantString);
-    fontLoader.addFont(byteData as Future<ByteData>);
-    await fontLoader.load();
-  }
+  if (byteData == null) return;
+  final fontData = await byteData;
+  if (fontData == null) return;
+
+  final fontLoader = FontLoader(familyWithVariantString);
+  fontLoader.addFont(byteData as Future<ByteData>);
```
Comment (pennzht) [inline lib/src/google_fonts_base.dart:189]: Should it be like this?
```dart
  fontLoader.addFont(Future.value(fontData));
```

Comment (kevmoo) [reply lib/src/google_fonts_base.dart:189]: I don't think so. What does this get us?

Comment (pennzht) [reply lib/src/google_fonts_base.dart:189]: https://github.com/material-foundation/google-fonts-flutter/issues/146#issuecomment-782176828

Because `addFont` requires a `Future<ByteData>` as input, and the value is already loaded, we can simply pack it inside a `Future.value` and pass it in.

Comment (kevmoo) [reply lib/src/google_fonts_base.dart:189]: Could it take `FutureOr<ByteData>`? Avoiding an unnecessary wrap would be nice.

Comment (kevmoo) [reply lib/src/google_fonts_base.dart:189]: Also, it seems this case isn't covered in tests, right?

Comment (pennzht) [reply lib/src/google_fonts_base.dart:189]: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/font_loader.dart#L36-L43

Unfortunately it takes only `Future<ByteData>`

Comment (kevmoo) [reply lib/src/google_fonts_base.dart:189]: The current impl just reuses the existing future, right? That should be fine – instead of creating a new one.

Comment (gaetschwartz) [reply lib/src/google_fonts_base.dart:189]: Exactly, the current uses the existing future.

Comment (kevmoo) [reply lib/src/google_fonts_base.dart:189]: @gaetschwartz I think you'll need to wrap the value. So you have a `Future<ByteData>` instead of `Future<ByteData?>`

Have you tested this workflow to make sure it works? Are there tests?

Comment (gaetschwartz) [reply lib/src/google_fonts_base.dart:189]: I ran the examples and it worked without but I can still do if you want

**Comment by kevmoo** [comment]
@gaetschwartz – we have zero tests here, it appears.

have you verified that your latest bits pass?

Should we add tests? Seems like a gnarly thing to let through w/out verification

**Comment by gaetschwartz** [comment]
Alright, can I make the method public with @vsisibleForTesting ? @kevmoo

**Comment by kevmoo** [comment]
> Alright, can I make the method public with @vsisibleForTesting ? @kevmoo

Sure!

**Comment by gaetschwartz** [comment]
@kevmoo After re-reading the code I don't really think we need to make a specific test as the only place where it is used is `loadFontIfNecessary` and this function is tested extensively. `loadFontIfNecessary` relies on it to load the appropriate fonts. 

TLDR; I don't think we need tests, thoughts ?

**Comment by kevmoo** [comment]
I defer to the package owners here, I guess. Just don't want this blowing up in production!

**Comment by gaetschwartz** [comment]
Understandably. Is it @pennzht ?

**Comment by pennzht** [comment]
@johnsonmh @clocksmith 

1. Should we add tests for this scenario?
2. If we should, should we add them in this PR?

**Comment by gaetschwartz** [comment]
@pennzht I don't really know how to mock `FontLoader` but what I can try is calling `_loadFontByteData` with different arguments and make sure it doesn't fail. Does it seem right to you ?

File: test/load_font_if_necessary_test.dart [hunk 2]
```diff
@@ -278,4 +278,25 @@ void main() {
...
+      () async => loadFontByteData('fontFamily', null),
+      returnsNormally,
+    );
+
+    expect(
+      () async => loadFontByteData('fontFamily',
+          Future.delayed(Duration(milliseconds: 100), () => null)),
+      returnsNormally,
+    );
+  });
```
Comment (gaetschwartz) [inline test/load_font_if_necessary_test.dart]: I added this test, thoughts ?

Comment (kevmoo) [reply test/load_font_if_necessary_test.dart]: Seems reasonable to me!

**Review by clocksmith — APPROVED** [review]

**Comment by kevmoo** [comment]
@clocksmith – would you merge?

**Review by pennzht — APPROVED** [review]

**Comment by pennzht** [comment]
@gaetschwartz Thank you so much for your PR!

**Comment by kevmoo** [comment]
@pennzht – are you going to merge?
